rfcs/0109-nixpkgs-generated-code-policy.md

```diff
@@ -0,0 +1,210 @@
+start-date: 2021-10-12
+author: John Ericson (@Ericson2314)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: @L-as @grahamc @sternenseemann
+shepherd-leader: @sternenseemann
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Nixpkgs contains non-trivial amounts of generated, rather than hand-written code.
+We want to start systematizing to make it easier to maintain.
+There is plenty of future work building upon this we could do, but we stop here for now to avoid needing to change any tools (Nix, Hydra, etc.).
+
+# Motivation
+[motivation]: #motivation
+
+Nixpkgs, along with every other distro, also faces a looming crisis: new open source software is increasingly not intended to be packaged by distros at all.
+Many languages now support very large library ecosystems, with dependencies expressed in a language-specific package manager.
+
+Right now, to deal with these packages, we either convert by hand, or commit lots of generated code into Nixpkgs.
+But I don't think either of those options is healthy or sustainable.
+The problem with the first is sheer effort; we'll never be able to keep up.
+The problem with the second is bloating Nixpkgs but more importantly reproducibility: If someone wants to update that generated code it is unclear how.
+All these mean that potential users coming from this new model of development find Nix / Nixpkgs cumbersome and unsuited to their needs.
+
+The lowest hanging fruit is to systematize our generated code.
+We should ensure anyone can update the generated code, which means it should be built in derivations not some ad-hoc way.
+In short, we should apply the same level of rigour that we do for packages themselves to generated code.
+
+# Detailed design
+[design]: #detailed-design
+
+## Nixpkgs
+
+1. Establish the policy that all generated code in nixpkgs must be produced by a derivation.
+   The derivation should be built by CI (so exposed as some Nixpkgs in some fashion).
+
+2. Implement script(s) for maintainers which automatically builds these derivations and vendors their results to the appropriate places.
+   Running such scripts should be sufficient to regenerated all generated code in Nixpkgs.
+
+   Greenfield tooling should not be merged unless it complies with the policy from day one.
+   Existing non-compliant tooling doesn't need to be ripped out of Nixpkgs, but the "grace period" in which is brought to compliance should be bounded.
+
+3. Ensure via CI that the vendored generated code is exactly what running the scripts produce.
+   This check should be one of the "channel blocking" CI jobs.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Impurities
+
+Many `lang2nix`-type tools have impure steps today.
+Since these tools must be only invoked inside the derivations to generate code, the impure inputs must be gotten via fixed output derivations.
```

Comment:
I would strongly recommend against Fixed-ouput drvs. they lead to a situation where it’s trivial to run into a reproducibility problem. The prefetching step as mentioned a few lines later is the way to go, since that way only one maintainer generates the file and we are not at the mercy of $version resolver being reproducible.

Reply:
Version solvers etc. should be executed inside the normal derivation. Fixed output derivations would be package repository indices etc.

Reply:
@Profpatsch I am a bit confused what you mean? I mean fixed output derivations to fetch things, not fixed output derivations that to a lot of arbitrary "vendoring" work.

Reply:
Okay, but then I’d just call them “fixed output fetchers” here, to make the distinction. 

Cause you can also used FODs for doing arbitrary stuff.

Reply:
I have a similar concern (perhaps it is the same concern).  Let me be very concrete and use Rust's `crates.io` as an example.

Rust's `cargo update` does a `git` fetch from the `master` branch of `https://github.com/rust-lang/crates.io-index`.  Right now `buildRustPackage` doesn't pin the revision of this repo -- it still fetches `master` and relies on the fact that the crates.io team never deletes `package:version` entries except when a package is "yanked" due to malware or vulnerabilities.

When crates.io yanks a package, people who use `cache.nixos.org` don't notice, because they hit the nixos cache, which never sees the yank.  People who build without any substituters will get a build failure.  I consider this to be a reproducibility problem.

@Ericson2314 / @sternenseeman: if we were to use `crate2nix`-generated nix code for Rust's `crates.io` (the way we currently use `cabal2nix`-generated code for Haskell's Hackage), this RFC says that code should be generated by a derivation.  Would that derivation fetch `https://github.com/rust-lang/crates.io-index` from a pinned revision or would it use the `master` branch?

Reply:
I have the same (I think) concern.

In order to be totally concrete I will use Rust as an example.

Suppose that we use `crate2nix`-generated nix code for Rust's `crates.io`, the way we currently use `cabal2nix`-generated code for Haskell's Hackage.  Suppose also that we stop using upstream's `Cargo.lock` files in order to be able to apply security updates to deep dependencies (e.g. `ring`, the Rust crypto library), to get better build/disk sharing, and to be able to drop the 74 `Cargo.lock` patches currently in nixpkgs.

In order to generate our own `Cargo.lock` files, the `crate2nix`-running derivation would need to run `cargo update`.  Rust's `cargo update` does a `git` fetch from the `master` branch of `https://github.com/rust-lang/crates.io-index`.  If we take this approach, when crates.io yanks a package, people who use `cache.nixos.org` don't notice, because they hit the nixos cache, which never sees the yank.  People who build without any substituters will get a build failure.

I consider this to be a reproducibility problem.

@Ericson2314 / @sternenseemann : Would that derivation fetch `https://github.com/rust-lang/crates.io-index` from the `master` branch, or would it fetch from a pinned git revision?  In the latter case, is this something we should require as a matter of policy?  If it is required as a matter of policy it should be rephrased in a way that isn't Rust/`crates.io`-specific of course.

Note: an earlier version of this comment contained incorrect statements about `buildRustPackage`.  If you saw that, please ignore it.

Reply:
> we are not at the mercy of $version resolver being reproducible.

Aren't we that anyway no matter how we implement it? Vendored FOD or not? We somehow need to pin dependency versions and that part must be reproducible.

Reply:
I would suggest fixing the wording to say something along the lines of "while making sure that the impure steps are as small as possible", albeit with better wording.

Reply:
How about this?
```suggestion
Since these tools must be only invoked inside the derivations to generate code, they must operate purely.
Any code that needs to be fetched must be done separately, e.g. with fixed-output derivations.
Any work deciding *what needs to be fetched* in the first place must be done impurely and out of band; this should be as little work as possible.
```

Reply:
There is the new "impure" derivation that might be useful here. If the CI checks for on-disk representation stability, this might help surface reproducibility problems. It also saves from downloading dependencies twice when updating FOD hashes.

Reply:
I don't think that is necessary, in most cases the input addressed approach will be ok: We have fixed-output data files as an input (which is updated via an "impure step" i.e. an update script) and run a solver in a regular derivation (which is hopefully deterministic!).

For taming quicklisp I don't think impure derivations are useful, since they are still sandboxed, aren't they?

Reply:
For Quicklisp we look at the packages themselves to check the dependency information (index contains some claims about dependencies that are usually correct but we have seen some exceptions so we don't trust that blindly)

I don't think Quicklisp has lockfiles: normal mode of operations works within a single globally coordinated (and tested) version set.

Reply:
@Ericson2314 I think your suggested change from above is good and should be applied.

The approach of running a version resolver inside a derivation is not a practical approach for some ecosystems.
The reason usually being that there is no package index readily available to feed into the `solver derivation`.

As mentioned above, some ecosystems expose dependency info only inside package sources, not offering a central index.

We can build our own package index by crawling all packages regularly, but the upstream resolver will most likely be incompatible to our custom package index.
-> we now have to invent our own dependency resolver as well.

I already went through all of that when building [mach-nix](https://github.com/DavHau/mach-nix) which is a nix code generator for python using a custom python package index [pypi-deps-db](https://github.com/DavHau/pypi-deps-db).

My personal experience is that this approach introduces significant maintenance overhead. Not only must the `crawler`/`db`/`resolver` be compatible to the complete history of standardization of the ecosystem but also adopt all upcoming standards to not fall out of sync.

Therefore think it is a good idea for the RFC to allow a minimal amount of `impure operations`. This simply allows to use upstream tooling to lock down the inputs.